### PR TITLE
Many parser should not handle infinite loop

### PR DESCRIPTION
**Pull request by minamorl**
Related to #197 

This changes contains runtime detection of infinite loop. If current parsing position is exact same as each parser's result, it should be infinite loop.

### CODE REVIEW

**Comment by wavebeem** [comment]
I really like this! It could be backwards incompatible though, if someone uses a parser with side effects in it. I think maybe it makes more sense to wait for Parsimmon 2.0 to accept this PR?

Hmm.

**Comment by minamorl** [comment]
As you pointed out, there's a possibility to break backward compatibility. All context-dependent parsers, which decides to consume input by context or not, will be affected by this change. It means the parser switches context inside infinite loop. Such kind of parser is too dependent to current `many()` implementation.

FYI I researched other parsing library [Nom(Rust)](https://github.com/Geal/nom) and [Parserc(Haskell)](https://hackage.haskell.org/package/parsec-3.1.11). Both of these reject empty string. Here's a reference:
- https://github.com/Geal/nom/issues/27
- https://hackage.haskell.org/package/parsec-3.1.11/docs/src/Text.Parsec.Prim.html#manyErr

My opinion is it is okay to ship this under current 1.x version.

**Comment by wavebeem** [comment]
Hmm, I think you're right. I should clarify in https://github.com/jneen/parsimmon/blob/master/API.md#side-effects that side effects also include randomness or reading from global state during parsing. It's already an expectation that bad things can happen if you do that.

File: test/core/many.test.js [hunk 1]
```diff
@@ -18,4 +18,9 @@ suite('many', function() {
     assert.equal(parser.parse('xxxxxy').value, 'y');
   });
 
+  test('can process infinite loop', function() {
```
Comment (wavebeem) [inline test/core/many.test.js]: `test('aborts on 0-character parse', ...)` maybe?

File: src/parsimmon.js [hunk 2]
```diff
@@ -396,14 +396,16 @@ _.then = function(next) {
 
 _.many = function() {
   var self = this;
-
   return Parsimmon(function(input, i) {
     var accum = [];
     var result = undefined;
-
+
```
Comment (wavebeem) [inline src/parsimmon.js]: please remove the spaces on this line

File: src/parsimmon.js [hunk 3]
```diff
@@ -396,14 +396,16 @@ _.then = function(next) {
...
     var accum = [];
     var result = undefined;
-
+    
     for (;;) {
       result = mergeReplies(self._(input, i), result);
       if (result.status) {
+        if (i === result.index) {
+          return makeFailure(i, 'Infinite loop has been detected in many() parser');
+        }
```
Comment (wavebeem) [inline src/parsimmon.js]: hmm, should this use makeFailure, or `throw new Error(...)`. it isn't a failed parse? it's a broken parser? i don't know.

Comment (minamorl) [inline src/parsimmon.js]: Seems you are right. I will change to Error so that it is more suitable in this context.

Comment (minamorl) [inline src/parsimmon.js]: this is off topic but i wish if parsimmon had custom errors containing its own kind...

Comment (wavebeem) [inline src/parsimmon.js]: that might be good!

Comment (wavebeem) [reply src/parsimmon.js]: @minamorl Do you still have time to change to `throw new Error(...)`?

**Comment by wavebeem** [comment]
@minamorl do you still want to work on this PR? or should I finish it

**Comment by wavebeem** [comment]
Closing in favor of https://github.com/jneen/parsimmon/pull/221 due to merge conflicts. Thank you @minamorl for your work on this issue!

**Pull request closed without merging**